kernel/libs/page_allocator/src/buddy_allocator.rs

```diff
@@ -0,0 +1,236 @@
+                // Split buffers
+                for j in (class + 1..i + 1).rev() {
+                    if let Some(block_ref) = self.free_list[j].iter().next() {
+                        let block = *block_ref;
+                        self.free_list[j - 1].insert(block + (1 << (j - 1)));
+                        self.free_list[j - 1].insert(block);
+                        self.free_list[j].remove(&block);
+                    } else {
+                        return None;
+                    }
+                }
+
+                let result = self.free_list[class].iter().next();
+                if let Some(result_ref) = result {
+                    let result = *result_ref;
+                    self.free_list[class].remove(&result);
+                    self.allocated += 1 << class;
+                    return Some(result);
+                } else {
+                    return None;
+                }
+            }
+        }
+        None
+    }
+
+    /// Deallocate a range of frames [frame, frame+count) from the frame
+    /// allocator.
+    ///
+    /// The range should be exactly the same when it was allocated, as in heap
+    /// allocator.
+    ///
+    /// # Safety
+    ///
+    /// Do not deallocate the same range twice.
+    pub fn dealloc(&mut self, start_frame: usize, count: usize) {
+        self.dealloc_power_of_two(start_frame, count.next_power_of_two().trailing_zeros())
+    }
+
+    /// Deallocate a range of frames described by count's power of two from the
+    /// allocator.
+    fn dealloc_power_of_two(&mut self, start_frame: usize, power: u32) {
+        let class = power as usize;
+
+        // Merge free buddy lists
+        let mut current_ptr = start_frame;
+        let mut current_class = class;
+        while current_class < self.free_list.len() {
+            let buddy = current_ptr ^ (1 << current_class);
+            if self.free_list[current_class].remove(&buddy) {
+                // Free buddy found
+                current_ptr = min(current_ptr, buddy);
+                current_class += 1;
+            } else {
+                self.free_list[current_class].insert(current_ptr);
+                break;
+            }
+        }
+
+        self.allocated -= 1 << class;
+    }
+
+    /// Given frames, described by a range of **frame number** [start, end),
+    /// mark them as allocated. Make sure they can be correctly deallocated
+    /// afterwards, while will not be allocated before deallocation.
+    ///
+    /// # Panics
+    ///
+    /// The function panics if no suitable block found for the given range.
+    ///
+    /// # TODO
+    ///
+    /// We are working on a more efficient implementation, based on free meta
+    /// (i.e., current unused meta) and in the format of linked list. By
+    /// introducing the free meta, we can reduce the time complexity of this
+    /// function from O(log(n)) to O(n).
+    pub fn alloc_specific(&mut self, start: usize, end: usize) {
```

Comment:
Do not write `TODO` in the documentation.

Besides, when the linked list is introduced we can only do it `O(n)`, `O(n)` to `O(n)` is not a reduction.